File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/pure/Mailbox.kt
// ── spine/pure/mailbox — what a barge-in consumer can be told (12.1–12.3) ──
// Three kinds of message, and the parent declares `source` ONCE so every one of
// them says where it came from. That single property is doing three jobs: it is the
// key conflation counts by, the scope dedupe runs in, and the attribution an
// Interrupt or a Drain carries onto the timeline.
//
// The three policies are the three variants, not three flags on one:
//
//     Input      reason about this          — conflate or queue, per the source's policy
//     Interrupt  PREEMPT                    — cancel the running turn, JOIN it, then run
//     Drain      DEFER, then finalize       — never preempts, never cancels
//
// ZERO I/O: no channel, no coroutine, no clock. The transport is pure; the machinery
// that moves it lives in spine/concurrency (gate check C8).

package adr.spine.pure

/** One thing the consumer can be told. Closed: there is no fourth kind of barge-in. */
sealed class Message(
    /** Where this came from. Declared once, carried by every variant (G12). */
    open val source: SourceName,
) {

    /**
     * A stimulus to reason about.
     *
     * `staged` is the NARROW variant — `StagedInput.Perceived`, never `StagedInput`.
     * `Recalled` has exactly ONE production site in the system, the consumer itself
     * (mirroring C7's one-production-site rule for `ToolResult`), so a producer
     * cannot post a forged relay snapshot into a turn.
     *
     * The dedupe key is `staged.key` — ONE field, and it lives on the value the step
     * COMMITS, so the envelope can never dedupe on one key while the record pins
     * another, and a restarted consumer rebuilds its dedupe scope from the timeline
     * alone. Non-null always: the durable policy dedupes on it, the perishable policy
     * ignores it. No nullable field, no branch, no "which mode am I in?".
     */
    data class Input(
        override val source: SourceName,
        val staged: StagedInput.Perceived,
    ) : Message(source)

    /** PREEMPT: the running turn is cancelled and JOINED before this one starts (12.3). */
    data class Interrupt(override val source: SourceName, val reason: String) : Message(source)

    /** DEFER: wait for the running turn, then finalize and stop. Never preempts (12.3). */
    data class Drain(override val source: SourceName, val reason: String) : Message(source)
}

/**
 * How a source's inputs behave when the consumer is busy. A CLOSED CHOICE, per
 * SOURCE — not a boolean and not an app-wide mode, because 12.2's note is about the
 * source's nature and a real deployment routinely has a durable ticket queue and a
 * perishable sensor feed at the same time.
 *
 * **The default is [DurableQueue]** ([policyFor] below). Three reasons, in order:
 *
 *  1. 12.2 says newest-input-wins is "exactly wrong for a durable work queue (the
 *     flagship server-agent deployment)". A default that is wrong for the flagship
 *     deployment is the wrong default.
 *  2. The failure modes are asymmetric. Durable-on-perishable processes stale items:
 *     wasteful, observable, recoverable. Perishable-on-durable SILENTLY LOSES WORK.
 *  3. It matches 14.3's posture: the safe classification is the default and the
 *     cheap one has to be declared and reviewed — exactly what `Verb` already does
 *     for reversibility.
 */
sealed class InputPolicy(
    /** The source this policy governs. Declared once (G12). */
    open val source: SourceName,
) {

    /** Newest-input-wins: while busy, conflate to the latest and COUNT the drops (12.2). */
    data class Perishable(override val source: SourceName) : InputPolicy(source)

    /** Never conflate: dedupe on the source key and ack only after the commit (12.2). */
    data class DurableQueue(override val source: SourceName) : InputPolicy(source)
}

/**
 * The per-source policy table. The LIST is the same on every lookup inside one
 * consumer, so it is constructor state and drops out of the call — the split the
 * boundary's ActionResolution already makes. An unlisted source gets
 * [InputPolicy.DurableQueue]: documented, not silent.
 */
class InputPolicies(private val policies: List<InputPolicy>) {
    fun forSource(source: SourceName): InputPolicy =
        policies.firstOrNull { it.source == source } ?: InputPolicy.DurableQueue(source)
}

/**
 * How long a cancel may take before the consumer stops waiting on it.
 *
 * 12.3's honest caveat is that an unbounded join is exactly a hang, so the bound is
 * real — and what preserves the no-interleave guarantee when it blows is the
 * REVOCATION LATCH in spine/concurrency/consumer, not the join.
 */
const val CANCEL_DEADLINE_MS: Millis = 200

/** The same bound for a Drain's defer. On timeout it revokes, reports and finalizes anyway. */
const val DRAIN_DEADLINE_MS: Millis = 1_000

File: agent-driven-architecture-main/examples/typescript/src/spine/pure/mailbox.ts
// ── spine/pure/mailbox — what can arrive while a turn is running (12.1) ─────
// ZERO I/O. Three things can be posted to a running agent, and they demand
// three DIFFERENT policies, which is why they are three variants of one sealed
// set rather than one message with a flag:
//
//   Input      a stimulus to reason about        → conflate or queue (§ policy)
//   Interrupt  "stop what you are doing"         → PREEMPT
//   Drain      "finish, then stop"               → DEFER, then finalize
//
// The parent declares `source` ONCE. Every message says where it came from,
// because the source is the key conflation counts by, the scope dedupe runs in,
// and the attribution an Interrupt or a Drain carries.

import type { SourceName } from "./ids";
import type { Perceived } from "./staged";

export type MessageKind = "Input" | "Interrupt" | "Drain";

export interface MessageBase {
  readonly kind: MessageKind;
  readonly source: SourceName;
}

/** A stimulus to reason about.
 *
 *  `staged` is typed `Perceived` — the NARROW variant, not `StagedInput`.
 *  `Recalled` has exactly one production site in the system (the consumer), so
 *  a producer cannot post a forged relay snapshot through the mailbox.
 *
 *  The dedupe key is `staged.key` — ONE field, and it lives on the value the
 *  step COMMITS, so the envelope can never dedupe on one key while the record
 *  pins another, and a restarted consumer rebuilds its dedupe scope from the
 *  timeline alone. Non-null always: the durable policy dedupes on it, the
 *  perishable policy ignores it. No nullable, no branch. */
export interface InputMessage extends MessageBase {
  readonly kind: "Input";
  readonly staged: Perceived;
}

/** PREEMPT: cancel the in-flight turn, join it, then reason about this. */
export interface InterruptMessage extends MessageBase {
  readonly kind: "Interrupt";
  readonly reason: string;
}

/** DEFER: never preempts. Wait for the running turn, then finalize, then stop. */
export interface DrainMessage extends MessageBase {
  readonly kind: "Drain";
  readonly reason: string;
}

export type Message = InputMessage | InterruptMessage | DrainMessage;

/** The discriminant as a CONSTANT, and the ONE place it is compared. A literal at
 *  every call site is a typo away from a silent false; a named guard is not. */
export const INPUT_KIND = "Input" as const;

export function isInput(message: Message): message is InputMessage {
  return message.kind === INPUT_KIND;
}

export function input(source: SourceName, staged: Perceived): InputMessage {
  return { kind: "Input", source, staged };
}

export function interrupt(source: SourceName, reason: string): InterruptMessage {
  return { kind: "Interrupt", source, reason };
}

export function drain(source: SourceName, reason: string): DrainMessage {
  return { kind: "Drain", source, reason };
}

// ── The two input policies — a CLOSED CHOICE, scoped PER SOURCE ─────────────
// Not a boolean and not a flag: newest-input-wins and never-lose-an-item are
// different contracts, and a deployment routinely has a durable ticket queue and
// a perishable sensor feed at the same time (12.2). So the policy belongs to the
// SOURCE, and the consumer is constructed with a table.
//
// THE DEFAULT IS `DurableQueue`. Conflation is opt-in, per source, always
// declared. Three reasons, in order of weight:
//
//   1. 12.2 says newest-input-wins is "exactly wrong for a durable work queue",
//      which is the flagship server-agent deployment. A default that is wrong
//      for the flagship deployment is the wrong default.
//   2. The failure modes are ASYMMETRIC. Durable-on-perishable processes stale
//      items: wasteful, observable, recoverable. Perishable-on-durable SILENTLY
//      LOSES WORK: unrecoverable, and 14.4's own catalog calls that data loss.
//   3. It matches the architecture's posture. 14.3's default-deny makes the safe
//      classification the default and forces the cheap one to be declared and
//      reviewed. `Verb` already does this; `InputPolicy` does it the same way.

export type InputPolicyKind = "Perishable" | "DurableQueue";

export interface InputPolicyBase {
  readonly kind: InputPolicyKind;
  readonly source: SourceName;
}

/** Newest-input-wins. A busy-drop conflates to the latest and FOLDS A COUNT —
 *  observable, never silent (12.2). */
export interface Perishable extends InputPolicyBase {
  readonly kind: "Perishable";
}

/** Never conflates. Dedupes on `key`, and acks only AFTER the commit, so a
 *  crash before the ack re-delivers rather than loses (12.2). */
export interface DurableQueue extends InputPolicyBase {
  readonly kind: "DurableQueue";
}

export type InputPolicy = Perishable | DurableQueue;

export function perishable(source: SourceName): Perishable {
  return { kind: "Perishable", source };
}

export function durableQueue(source: SourceName): DurableQueue {
  return { kind: "DurableQueue", source };
}

/** An unlisted source gets `DurableQueue`. Documented, not silent. */
export function policyFor(table: readonly InputPolicy[], source: SourceName): InputPolicy {
  return table.find((p) => p.source === source) ?? durableQueue(source);
}

// ── The two bounds ──────────────────────────────────────────────────────────
// 12.3's own honest caveat is that an unbounded join is exactly a hang. These
// make the bound real. Both are injectable at construction.

/** How long a PREEMPT waits for the cancelled turn before abandoning it. */
export const CANCEL_DEADLINE_MS = 250;

/** How long a DRAIN waits for the running turn before finalizing anyway. */
export const DRAIN_DEADLINE_MS = 1000;
